### PR TITLE
Add `no-new-buffer` rule - fixes #59

### LINKED ISSUE #59: Enforce `Buffer.from` instead of `new Buffer`
Since `new Buffer()` is being deprecated since Node.js >= 6, it might be nice if we could enforce this some day in the future. Not sure if it makes sense to already create this rule because ESLint can't target specific Node.js versions, right?

### PR DESCRIPTION
**Pull request by SamVerschueren**
:tada: Feedback more then welcome!

I was struggling with using `fixer.replaceText` and thought I had to parse the arguments myself untill I found out about `fixer.replaceTextRange` :).

### CODE REVIEW

File: docs/rules/no-new-buffer.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# Enforce the use of `Buffer.from()` and `Buffer.alloc()` instead of the deprecated `new Buffer()`
+
+Enforces the use of [Buffer.from](https://nodejs.org/api/buffer.html#buffer_class_method_buffer_from_array) and [Buffer.alloc()](https://nodejs.org/api/buffer.html#buffer_class_method_buffer_alloc_size_fill_encoding) instead of [new Buffer()](https://nodejs.org/api/buffer.html#buffer_new_buffer_array) which has been deprecated.
```
Comment (sindresorhus) [inline docs/rules/no-new-buffer.md]: Should there be a comma before `which has been deprecated.`?

File: docs/rules/no-new-buffer.md [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
```
Comment (sindresorhus) [inline docs/rules/no-new-buffer.md]: Maybe we should be explicit about those two methods being available in Node.js 4 and up.

File: readme.md [hunk 3]
```diff
@@ -55,7 +55,7 @@ Configure it in `package.json`.
 - [no-process-exit](docs/rules/no-process-exit.md) - Disallow `process.exit()`.
 - [throw-new-error](docs/rules/throw-new-error.md) - Require `new` when throwing an error. *(fixable)*
 - [number-literal-case](docs/rules/number-literal-case.md) - Enforce lowercase identifier and uppercase value for number literals. *(fixable)*
-
```
Comment (sindresorhus) [inline readme.md]: Don't remove the newline.

File: readme.md [hunk 4]
```diff
@@ -55,7 +55,7 @@ Configure it in `package.json`.
...
+- [no-new-buffer](docs/rules/no-new-buffer.md) - Enforce the use of `Buffer.from()` and `Buffer.alloc()` instead of the deprecated `new Buffer()`
```
Comment (sindresorhus) [inline readme.md]: Needs to be marked as fixable.

File: rules/no-new-buffer.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+			if (node.callee.name === 'Buffer') {
+				const method = inferMethod(node.arguments);
+				const range = [
+					node.start,
+					node.callee.end
+				];
+
+				context.report({
+					node,
+					message: `\`new Buffer()\` is being deprecated, use \`Buffer.${method}()\` instead`,
```
Comment (sindresorhus) [inline rules/no-new-buffer.js]: Drop the `is being` part. It *is* deprecated.

File: rules/no-new-buffer.js [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+			}
+		}
+	};
+};
+
+module.exports = {
+	meta: {
+		fixable: 'code'
+	},
+	create
```
Comment (sindresorhus) [inline rules/no-new-buffer.js]: Can you put `create` at the top to be consistent with the other rules?

File: rules/no-new-buffer.js [hunk 7]
```diff
@@ -0,0 +1,35 @@
+'use strict';
+const inferMethod = args => {
+	if (args.length > 0 && typeof args[0].value === 'number') {
+		return 'alloc';
+	}
+
+	return 'from';
+};
```
Comment (sindresorhus) [inline rules/no-new-buffer.js]: This could be a one-liner.

File: docs/rules/no-new-buffer.md (hunk 1 above)
Comment (SamVerschueren) [inline docs/rules/no-new-buffer.md]: Not sure, which one do you think sounds better :D?

File: docs/rules/no-new-buffer.md (hunk 2 above)
Comment (SamVerschueren) [inline docs/rules/no-new-buffer.md]: Good point btw, people should be well informed.

File: docs/rules/no-new-buffer.md (hunk 1 above)
Comment (SamVerschueren) [inline docs/rules/no-new-buffer.md]: With the new sentence `which has been deprecated since Node.js 4`, I believe a comma makes sense.

**Comment by sindresorhus** [comment]
@SamVerschueren Can you fix the merge conflict?

**Comment by kevva** [comment]
Nice work with this one!

File: docs/rules/no-new-buffer.md [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+
+## Fail
+
+```js
+const buf = new Buffer('7468697320697320612074c3a97374', 'hex');
+const buf = new Buffer([0x62, 0x75, 0x66, 0x66, 0x65, 0x72]);
+```
+
+```js
```
Comment (jfmengels) [inline docs/rules/no-new-buffer.md]: Why the two distinct code blocks?

**Comment by SamVerschueren** [comment]
The [first invalid](https://github.com/sindresorhus/eslint-plugin-unicorn/pull/61/files#diff-bdd2d7fa0e7a33141bf1a62d561e2fc8R33) test is a test with `new Buffer()`

And the two last invalid tests use an identifier. Or maybe you mean something else?

File: docs/rules/no-new-buffer.md (hunk 8 above)
Comment (SamVerschueren) [inline docs/rules/no-new-buffer.md]: Good question, I wanted to make a difference between `new Buffer` being converted to `Buffer.from` and the `new Buffer` being converted to `Buffer.alloc` as you can see in the `Pass` section below. To keep those blocks consistent, I splitted them up. I could merge them though, whatever you guys prefer.

**Comment by jfmengels** [comment]
Ah, my bad! LGTM then.

> But the fix should probably not fix if it can't determine the right method to use, otherwise you could inject bugs in the codebase.

Still wondering about this. But I leave that to someone who knows it better.

**Comment by SamVerschueren** [comment]
> But the fix should probably not fix if it can't know for sure, otherwise you could inject bugs in the codebase.

You mean `new Buffer()` to `Buffer.from()` which is actually invalid? I agree, this is an edge-case and I went for converting it. `new Buffer()` is invalid as well though. If someone else has a strong opinion on this, feel free to chime in :).

**Comment by jfmengels** [comment]
> You mean new Buffer() to Buffer.from() which is actually invalid?

I mean this

```js
let value = 10;
new Buffer(value) // should be fixed to Buffer.alloc
value = 'foo';
new Buffer(value) // should be fixed to Buffer.from
```
With most expressions or identifiers, you can't know for sure which fix to apply. Maybe not apply a fix when you can't determine that, and advise to use either `alloc` or `from`.

**Comment by SamVerschueren** [comment]
Darn, good catch. Even haven't though about that. Maybe not fixing that might be better indeed... I will look at that.

**Comment by sindresorhus** [comment]
> Maybe not apply a fix when you can't determine that, and advise to use either alloc or from.

I think it's safe to always assume `Buffer.from()` as even if we guess incorrectly, `const foo = 10; Buffer.from(foo);` would throw, as `Buffer.from()` doesn't accept a number. In addition, I've never seen anyone use a number there, so very much and edge-case.

> Welcome to the world of I wish JS had types so that I could know the type of X statically :D

Flow could have easily inferred it in this case without any type info. In most cases actually, it's ESLint being naive, not lack of types that is the show stopper.

**Comment by jfmengels** [comment]
> Flow could have easily inferred it in this case without any type info. In most cases actually, it's ESLint being naive, not lack of types that is the show stopper.

It probably could have in most cases yes. But you're right, ESLint could infer stuff on its own.

**Comment by sindresorhus** [comment]
So, are you guys good with merging this as is?

**Comment by SamVerschueren** [comment]
I'm ok. If someone complains, I could always try to track down the value of the reference by traversing the tree. But that might be a little bit overkill for now.

**Comment by jfmengels** [comment]
I think it's dangerous to apply the fix whenever, but if you think it's a sufficient edge case, and that people will notice it soon enough, let's merge this. But I have a bad gut feeling about doing so TBH.

**Comment by SamVerschueren** [comment]
Should I look into traversing the tree and try to infer the type ourselves? Could be easily extracted in a separate library. If the type could not be inferred (easily), we do not apply the fix, if it could be inferred, we apply it. Not sure if possible though, but could be very useful for other rules if we had a basic type inference system.

**Comment by sindresorhus** [comment]
> I think it's dangerous to apply the fix whenever

You're right, but I don't think this is practically going to affect anyone. Only very few libraries allocates the buffer size, and those will get an error about incorrect usage from `Buffer.from()` anyways, so what is really the harm other than a slightly incorrect transformation? It won't introduce any hidden bugs. We have to weigh complexity against risk, and I don't think the risk outweighs the complexity here.

**Comment by jfmengels** [comment]
If you feel that way, let's do it.

**Comment by jfmengels** [comment]
Thanks again @SamVerschueren :D

**Comment by sindresorhus** [comment]
I still think adding better type inference is a good idea. Especially if we can make something reusable as many rules would be better with some light type inference. Type inference could also increase the chances here of guessing the correct type, and making @jfmengels a little bit more at ease.

**Comment by sindresorhus** [comment]
Thanks @SamVerschueren. Great as always :)

**Comment by SamVerschueren** [comment]
Nice, thanks for all the feedback guys! I will add it to my todo list :).

**Pull request merged**